### PR TITLE
chore: adds protected endpoint for DB metrics

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

- Enables [Prisma's experimental metrics feature](https://www.prisma.io/docs/orm/prisma-client/observability-and-logging/metrics#about-metrics)
- Adds a protected endpoint that allow us to probe our servers to check DB related metrics
- This will allow us to adjust our pooling servers to prevent dropped connections due to timeouts

## Type of change

- Chore (refactoring code, technical debt, workflow improvements)
- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Locally send a GET request to `http://localhost:3000/api/metrics?apiKey=DUMMY_API_KEY`
- You should receive a metrics object:
    ```
    {
        "counters": [
            {
                "key": "prisma_client_queries_total",
                "labels": {},
                "value": 0,
                "description": "The total number of Prisma Client queries executed"
            },
            {
                "key": "prisma_datasource_queries_total",
                "labels": {},
                "value": 0,
                "description": "The total number of datasource queries executed"
            },
            {
                "key": "prisma_pool_connections_closed_total",
                "labels": {},
                "value": 0,
                "description": "The total number of pool connections closed"
            },
            {
                "key": "prisma_pool_connections_opened_total",
                "labels": {},
                "value": 1,
                "description": "The total number of pool connections opened"
            }
        ],
        "gauges": [
            {
                "key": "prisma_client_queries_active",
                "labels": {},
                "value": 0,
                "description": "The number of currently active Prisma Client queries"
            },
            {
                "key": "prisma_client_queries_wait",
                "labels": {},
                "value": 0,
                "description": "The number of datasource queries currently waiting for an free connection"
            },
            {
                "key": "prisma_pool_connections_busy",
                "labels": {},
                "value": 0,
                "description": "The number of pool connections currently executing datasource queries"
            },
            {
                "key": "prisma_pool_connections_idle",
                "labels": {},
                "value": 21,
                "description": "The number of pool connections that are not busy running a query"
            },
            {
                "key": "prisma_pool_connections_open",
                "labels": {},
                "value": 1,
                "description": "The number of pool connections currently open"
            }
        ],
        "histograms": [
            {
                "key": "prisma_client_queries_wait_histogram_ms",
                "labels": {},
                "value": {
                    "buckets": [
                        [
                            0,
                            0
                        ],
                        [
                            1,
                            1
                        ],
                        [
                            5,
                            0
                        ],
                        [
                            10,
                            0
                        ],
                        [
                            50,
                            0
                        ],
                        [
                            100,
                            0
                        ],
                        [
                            500,
                            0
                        ],
                        [
                            1000,
                            0
                        ],
                        [
                            5000,
                            0
                        ],
                        [
                            50000,
                            0
                        ]
                    ],
                    "sum": 0.0007080000000000001,
                    "count": 1
                },
                "description": "The distribution of the time all datasource queries spent waiting for a free connection"
            }
        ]
    }
    ```

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/prisma/schema.prisma [hunk 1]
```diff
@@ -9,7 +9,7 @@ datasource db {
 
 generator client {
   provider        = "prisma-client-js"
-  previewFeatures = ["views"]
+  previewFeatures = ["views", "metrics"]
```
Comment (keithwillcode) [inline packages/prisma/schema.prisma:12]: Should we put this behind an env variable and have it turned off by default? Might affect perf

Comment (zomars) [reply packages/prisma/schema.prisma:12]: I'm not too sure if we can do that for enabling preview features.

**Comment by keithwillcode** [comment]
@zomars Now that we have upgraded our prod DB and we also get similar stats directly from it, do we still want to add this?

File: packages/lib/server/rateLimit.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import type { NextApiHandler } from "next";
+
+import { HttpError } from "../http-error";
+import { rateLimiter } from "../rateLimit";
+
+export const rateLimitHandler: NextApiHandler = async (req) => {
```
Comment (keithwillcode) [inline packages/lib/server/rateLimit.ts:6]: Why did we need to build this rate limiting logic? Don't we already have this and the calculations are done for us?

**Comment by emrysal** [comment]
The rate limiting logic is a bit duplicate with existing rate limiting, at the moment drafting this PR - happy to merge only metrics if confirmed that the db performance isn't impacted - but honestly we have plenty of ways to track DB metrics and soon with Sentry Performance we have even more options.

**Comment by keithwillcode** [comment]
Going to close for now until we determine we really want/need this since we have other ways to tracking db perf.

**Pull request closed without merging**